fee: Round up fee calculation to avoid a lower than expected feerate

achow101 opened the pull request:
When calculating the fee for a feerate, it is possible that the final calculation will have fractional satoshis. Currently those are ignored via truncation which results in the absolute fee being rounded down. Rounding down is problematic because it results in a feerate that is slightly lower than the feerate represented by the `CFeeRate` object. A slightly lower feerate particularly causes issues for coin selection as it can trigger an assertion error. To avoid potentially underpaying the feerate (and the assertion), always round up the calculated fee.

A test is added for the assertion, along with a comment explaining what happens.

It is unlikely that a user can trigger this as it requires a very specific set of rounding errors to occur as well as the transaction not needing any change and being right on the lower bound of the exact match window. However I was able to trigger the assertion while running coin selection simulations, albeit after thousands of transactions and with some weird feerates.

achow101 committed:
fees: Always round up fee calculated from a feerate

When calculating the fee for a given tx size from a fee rate, we should
always round up to the next satoshi. Otherwise, if we round down (via
truncation), the calculated fee may result in a fee with a feerate
slightly less than targeted.

This is particularly important for coin selection as a slightly lower
feerate than expected can result in a variety of issues.

achow101 committed:
tests: Test for assertion when feerate is rounded down

When calculating a txs absolute fee, if the fee is rounded down to the
nearest satoshi, it is possible for the coin selection algorithms to
undercalculate the fee needed. This can lead to an assertion error in
some situations. One such scenario is added to
rpc_fundrawtransaction.py.

bitcoin-public-domain commented:
Is this another attempts to change the protocol to serve your or Blockstream's interest? 

**The fess are just fine how they are.** This is just a stupid pr which shouldn't be merged.

bitcoin-public-domain commented:
> A slightly lower feerate particularly causes issues for coin selection as it can trigger an assertion error.

You also doesn't made it clear what coin selection you talking about. I never had any problem with selecting coins or spending them. 

Seems to me you working on something at Blockstream which falls outside of the scope of the Bitcoin repository. Don't take it as an insult.

1 Bitcoin equals to 100,000,000 satoshi. **Paying 1 satoshi less or more for the transaction doesn't make any difference!** 

```cpp
/** Amount in satoshis (Can be negative) */
typedef int64_t CAmount;
 
static constexpr CAmount COIN = 100000000;
```

bitcoin-public-domain commented:
Doesn't make any difference in rounding down or up the fee. Paying one satoshi more or less doesn't even matter at all since a satoshi is basically worthless. 

> It is unlikely that a user can trigger this as it requires a very specific set of rounding errors to occur as well as the transaction not needing any change and being right on the lower bound of the exact match window. However I was able to trigger the assertion while running coin selection simulations, albeit after thousands of transactions and with some weird feerates.

This pr shouldn't be merged. Even op acknowledged that the end user is unlikely to face any issue like he did in his simulations.

darosior reviewed (commented):
Modifying `CFeeRate::GetFee` is a policy change. Increasing its return value (for some input) is potentially a policy rule tightening which may have security implications for applications with pre-signed transactions.

On the top of my head, i don't know of of an application for which it would be fatal. But could we instead go the safer path of containing the roundup in the wallet code rather than modifying `CFeeRate::GetFee`?

achow101 commented:
It seems odd to me that the wallet and policy should have different fee calculation results for the same transaction.

In terms of pre-signed transactions, what security implications do you imagine? I don't think it is possible for this to cause a transaction's feerate to fall below the default minimum relay fee because the result of the fee calculation at the minimum relay fee is always an integer since the minimum relay fee is an integer. I believe the same is true for the incremental relay fee and how that calculation works.

darosior commented:
>  I believe the same is true for the incremental relay fee and how that calculation works.

I don't think it's true in all cases. Take clawback tx A with fee such as it would be rounded up after this PR, not before. The application wants to replace it with clawback tx B with fee equal to `A's fee + incremental relay fee` such as it would not be rounded up.
Before this PR tx B would be accepted, after it would be refused.

Of course it'd be reckless for such an application to not take some leeway in addition to the incremental fee here, but still this PR would be a behaviour change for this edge case.

achow101 commented:
> I don't think it's true in all cases. Take clawback tx A with fee such as it would be rounded up after this PR, not before. The application wants to replace it with clawback tx B with fee equal to `A's fee + incremental relay fee` such as it would not be rounded up.
> Before this PR tx B would be accepted, after it would be refused.

How so? Can you provide a example? I can't think of how that would happen. `CFeeRate::GetFee` is never called on the original feerate in validation, and the calculation of a feerate for a transaction is not changed by this PR.

Consider the example I give in the test: tx A is a 1-in-1-out 110 vbyte tx that pays a feerate of 1.85 sat/vbyte for an absolute fee if 203.5 sats. With rounding down, A will pay 203 sats. We replace this with tx B where the only thing different is the output, but it is of the same type so the size remains the same. Paying the incremental relay fee should be enough for B to replace A. The incremental feerate is 1 sat/vbyte, so regardless of rounding, it is always 110 sats. So B pays 313 sats.

When validating whether B is allowed to replace A, it needs to pass two fee related checks: [`PaysMoreThanConflicts`](https://github.com/bitcoin/bitcoin/blob/master/src/policy/rbf.cpp#L124) (called [here](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L785)) and [`PaysForRBF`](https://github.com/bitcoin/bitcoin/blob/master/src/policy/rbf.cpp#L150) (called [here](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L806)). `PaysMoreThanConflicts` computes the feerates for A and B and just checks that B's feerate is greater than A's. The rounding of computing the feerate given the fee paid and size **did not change**, so both before and after this PR, the calculation for A is `203 * 1000 / 110 = 1845`. The calculation for B is `313 * 1000 / 110 = 2845`. B's feerate is greater than A's, so this check passes, both before and after this PR. `PaysForRBF` checks that the fees for B is greater than A. Then it checks that the difference is not less than the incremental relay fee. Because the incremental relay fee is an integer, the calculation before and after this PR is the same - 110 sats.

At no point in this checking process is the original feerate of 1.85 sats/vbyte used because validation does not know that was the original targeted feerate. The only times where validation calls `CFeeRate::GetFee` is to calculate the fee at the minimum relay feerate and the incremental relay feerate, both of which have integers as their default, so their calculations do not change. So I don't see how this change would have any effect on whether transactions are accepted.

darosior commented:
You are right. I also can't think of how this change would be a policy rule tightening.

ccr781 commented:
***@***.***

chaiwatkaewmanee

ในวันที่ ส. 11 ก.ย. 2021 07:36 Andrew Chow ***@***.***>
เขียนว่า:

> When calculating the fee for a feerate, it is possible that the final
> calculation will have fractional satoshis. Currently those are ignored via
> truncation which results in the absolute fee being rounded down. Rounding
> down is problematic because it results in a feerate that is slightly lower
> than the feerate represented by the CFeeRate object. A slightly lower
> feerate particularly causes issues for coin selection as it can trigger an
> assertion error. To avoid potentially underpaying the feerate (and the
> assertion), always round up the calculated fee.
>
> A test is added for the assertion, along with a comment explaining what
> happens.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/bitcoin/bitcoin/pull/22949
> Commit Summary
>
>    - fees: Always round up fee calculated from a feerate
>    - tests: Test for assertion when feerate is rounded down
>
> File Changes
>
>    - *M* src/policy/feerate.cpp
>    <https://github.com/bitcoin/bitcoin/pull/22949/files#diff-cc74b9dc8da4346ef6f877eab0aafa44feeea22d672a6516fe2d4f2fd65a74cc>
>    (4)
>    - *M* test/functional/rpc_fundrawtransaction.py
>    <https://github.com/bitcoin/bitcoin/pull/22949/files#diff-729d780cd34193bdca2c44e9765fb47063bfbf88ba50e0864dad12d747033995>
>    (28)
>
> Patch Links:
>
>    - https://github.com/bitcoin/bitcoin/pull/22949.patch
>    - https://github.com/bitcoin/bitcoin/pull/22949.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/22949>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APA4O7SBZWJ3ENIVQPCRH43UBKQCTANCNFSM5D2K6V6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jonatack reviewed (commented):
Tentative concept ACK after looking at the callers of `GetFee()` (not `GetFeePerK`, which should be a pass-through). The new functional tests fail where expected on master.
```
$ test/functional/rpc_fundrawtransaction.py 

  File "/home/jon/projects/bitcoin/bitcoin/test/functional/rpc_fundrawtransaction.py", line 1055, in test_feerate_rounding
    assert_raises_rpc_error(-4, "Insufficient funds", w.fundrawtransaction, rawtx, {"fee_rate": 1.85})
AssertionError: Unexpected stderr bitcoind: wallet/spend.cpp:814: bool CreateTransactionInternal(CWallet &, const std::vector<CRecipient> &, CTransactionRef &, CAmount &, int &, bilingual_str &, const CCoinControl &, FeeCalculation &, bool): Assertion `coin_selection_params.m_subtract_fee_outputs || fee_needed <= change_and_fee - change_amount' failed. != 
```
```
$ test/functional/wallet_keypool.py

  File "/home/jon/projects/bitcoin/bitcoin/test/functional/wallet_keypool.py", line 184, in run_test
    assert_equal(res["fee"], Decimal("0.00009706"))
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 49, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(0.00009705 == 0.00009706)
```

MarcoFalke commented on src/policy/feerate.cpp:
```diff
@@ -22,7 +24,7 @@ CAmount CFeeRate::GetFee(uint32_t num_bytes) const
 {
```
Wouldn't it make sense to rename this from `GetFee` to `RoundFee` to clarify that this is not a getter, but a calculation?

At the very least the doxygen should be updated.

achow101 replied:
Updated doxygen. I think keeping it `GetFee` is fine.

MarcoFalke commented:
> You are right. I also can't think of how this change would be a policy rule tightening.

This changes policy for non-default fee rate settings.

achow101 commented on src/policy/feerate.cpp:
```diff
@@ -22,7 +24,7 @@ CAmount CFeeRate::GetFee(uint32_t num_bytes) const
     const int64_t nSize{num_bytes};
 
-    CAmount nFee = nSatoshisPerK * nSize / 1000;
+    CAmount nFee = std::ceil(nSatoshisPerK * nSize / 1000.0);
```
I added the explicit casting but not the bypass for now.

achow101 commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1026,6 +1027,33 @@ def do_fund_send(target):
+        # In the former case, the calculated needed fee is higher than the actual fee being paid, so an assertion is reached
+        # To test this does not happen, we subtract 202 sats from the input value. If working correctly, this should
+        # fail with insufficient funds rather than bitcoind asserting.
+        rawtx = w.createrawtransaction([], [{self.nodes[0].getnewaddress(address_type="bech32"): 1 - 0.00000202}])
```
Done

jonatack commented:
Code review ACK d9ac0cdd5b7829dd45e1dab46fa349d2ce04d5cd modulo policy considerations, per `git diff 053ec2b d9ac0cd`, previously checked at 053ec2b that the new functional tests fail where expected on master, changes since are a named cast, documentation, and named args in the test

ariard commented:
I agree, I don't think this change qualifies as a policy rule tightening.

IIUC, previously we had transaction A with absolute fee 203.5 sats. Currently, it's rounded down to 203 sats. After this PR, it will be rounded up to 204 sats.

Note, I think `GetFee` is called in few subsystems beyond validation and wallet. Such as for block construction (`addPackageTxs`) or transaction announcement selection (L4802, in `src/net_processing`). As the pair of transactions is equally affected, I don't think the comparison is altered. I.e if transaction A was 203.5 sats and transaction B was 204.5 sats, respectivelly rounded down to 203 sats and 204 sats, now they will be rounded up to 204 sats and 205 sats ? The comparison should hold. AFAICT, the proposed change doesn't have positive or negative impacts on transaction propagation or block building.

(Though if it's change the displayed absolute fee at the interface-level, I think a release note would be appropriate, a user could consume the now-rounded up result and do a comparison with a still-rounded down result from another software than Core in a way which is obscure to us ?)

meshcollider commented:
This should also fix test failures like this: https://cirrus-ci.com/task/5800722060017664?logs=ci#L3900

MarcoFalke commented:
Concept ACK. If someone requests a feerate of 1.23 sat/vB, with a tx vsize of 208, it shouldn't result in 255 fee, but 256. Otherwise, the feerate will be 1.2259... sat/vB.

I think you'll also need to adjust the test, otherwise they won't check this:

```
test/functional/test_framework/util.py:def satoshi_round(amount):

achow101 commented:
@MarcoFalke Not quite sure what you want me to adjust.

`satoshi_round` doesn't need any changes afaict.

MarcoFalke commented:
`assert_fee_amount`, which uses `satoshi_round` is used in the tests to replicate `GetFee` and check that it gives the same result.

You are changing `GetFee` (in my previous example https://github.com/bitcoin/bitcoin/pull/22949#issuecomment-932045979) to return 256 instead of 255, so the tests also needs to be adjusted to do the same.

achow101 committed:
tests: Calculate fees more similarly to CFeeRate::GetFee

Because of floating point precision issues, not all of the rounding done
is always correct. To fix this, the fee calculation for
assert_fee_amount is changed to better reflect how CFeeRate::GetFee does
it.

First the feerate is converted to an int representing sat/kvb. Then this
is multiplied by the transaction size, divivided by 1000, and rounded up
to the nearest sat. The result is then converted back to BTC (divided by
1e8) and then rounded down to the nearest sat to avoid precision errors.

ccr781 commented:
***@***.***

chaiwat

ในวันที่ ศ. 17 ก.ย. 2021 18:31 Jon Atack ***@***.***>
เขียนว่า:

> ***@***.**** commented on this pull request.
>
> Tentative concept ACK after looking at the callers of GetFee() (not
> GetFeePerK, which should be a pass-through). The new functional tests
> fail where expected on master.
>
> $ test/functional/rpc_fundrawtransaction.py
>
>
>
>   File "/home/jon/projects/bitcoin/bitcoin/test/functional/rpc_fundrawtransaction.py", line 1055, in test_feerate_rounding
>
>     assert_raises_rpc_error(-4, "Insufficient funds", w.fundrawtransaction, rawtx, {"fee_rate": 1.85})
>
> AssertionError: Unexpected stderr bitcoind: wallet/spend.cpp:814: bool CreateTransactionInternal(CWallet &, const std::vector<CRecipient> &, CTransactionRef &, CAmount &, int &, bilingual_str &, const CCoinControl &, FeeCalculation &, bool): Assertion `coin_selection_params.m_subtract_fee_outputs || fee_needed <= change_and_fee - change_amount' failed. !=
>
>
> $ test/functional/wallet_keypool.py
>
>
>
>   File "/home/jon/projects/bitcoin/bitcoin/test/functional/wallet_keypool.py", line 184, in run_test
>
>     assert_equal(res["fee"], Decimal("0.00009706"))
>
>   File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 49, in assert_equal
>
>     raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
>
> AssertionError: not(0.00009705 == 0.00009706)
>
>
> ------------------------------
>
> In src/policy/feerate.cpp
> <https://github.com/bitcoin/bitcoin/pull/22949#discussion_r710958422>:
>
> > @@ -22,7 +24,7 @@ CAmount CFeeRate::GetFee(uint32_t num_bytes) const
>
>  {
>
>      const int64_t nSize{num_bytes};
>
>
>
> -    CAmount nFee = nSatoshisPerK * nSize / 1000;
>
> +    CAmount nFee = std::ceil(nSatoshisPerK * nSize / 1000.0);
>
>
> ee9a74d
> <https://github.com/bitcoin/bitcoin/commit/ee9a74d99ce784b3570ae606eac4dd9627727ec1>
> a couple of ideas
>
>    - explicit, named cast
>
> ⬇️ Suggested change
>
> -    CAmount nFee = std::ceil(nSatoshisPerK * nSize / 1000.0);
>
> +    // Be explicit that we're converting from double to long (CAmount) here.
>
> +    // We've previously had bugs creep in from silent double->int conversion.
>
> +    CAmount nFee{static_cast<CAmount>(std::ceil(nSatoshisPerK * nSize / 1000.0))};
>
>
> ------------------------------
>
>    - bypass this calculation in the frequent GetFeePerK() case
>
>     if (nSize != 1000) {
>
>         CAmount nFee{static_cast<CAmount>(std::ceil(nSatoshisPerK * nSize / 1000.0))};
>
>     }
>
> ------------------------------
>
> In test/functional/rpc_fundrawtransaction.py
> <https://github.com/bitcoin/bitcoin/pull/22949#discussion_r710959094>:
>
> > +
>
> +        addr = w.getnewaddress(address_type="bech32")
>
> +        self.nodes[0].sendtoaddress(addr, 1)
>
> +        self.generate(self.nodes[0], 1)
>
> +        self.sync_all()
>
> +
>
> +        # A P2WPKH input costs 68 vbytes; With a single P2WPKH output, the rest of the tx is 42 vbytes for a total of 110 vbytes.
>
> +        # At a feerate of 1.85 sat/vb, the input will need a fee of 125.8 sats and the rest 77.7 sats
>
> +        # The entire tx fee should be 203.5 sats.
>
> +        # Coin selection rounds the fee individually instead of at the end (due to how CFeeRate::GetFee works).
>
> +        # If rounding down (which is the incorrect behavior), then the calculated fee will be 125 + 77 = 202.
>
> +        # If rounding up, then the calculated fee will be 126 + 78 = 204.
>
> +        # In the former case, the calculated needed fee is higher than the actual fee being paid, so an assertion is reached
>
> +        # To test this does not happen, we subtract 202 sats from the input value. If working correctly, this should
>
> +        # fail with insufficient funds rather than bitcoind asserting.
>
> +        rawtx = w.createrawtransaction([], [{self.nodes[0].getnewaddress(address_type="bech32"): 1 - 0.00000202}])
>
>
> 053ec2b
> <https://github.com/bitcoin/bitcoin/commit/053ec2b901bee174de3243302c113b4a91733241>
> nit
> ⬇️ Suggested change
>
> -        rawtx = w.createrawtransaction([], [{self.nodes[0].getnewaddress(address_type="bech32"): 1 - 0.00000202}])
>
> +        rawtx = w.createrawtransaction(inputs=[], outputs=[{self.nodes[0].getnewaddress(address_type="bech32"): 1 - 0.00000201}])
>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/22949#pullrequestreview-757331938>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APA4O7QPOSWJU3L3MRHTVOLUCMRI7ANCNFSM5D2K6V6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

ryanofsky reviewed (approved):
Code review ACK 80dc829be7f8c3914074b85bb4c125baba18cb2c

promag reviewed (commented):
Tested ACK 80dc829be7f8c3914074b85bb4c125baba18cb2c.

Funny thing, looking at the 1st commit in GH we (kind of) had round-up:
https://github.com/bitcoin/bitcoin/blob/e071a3f6c06f41068ad17134189a4ac3073ef76b/main.h#L509

Later the code was refactored in c6cb21d17ab8097b6a425d37e48c955fbb0e9f0c (#3959) and the round-up was dropped.

lsilva01 reviewed (approved):
tACK 80dc829

meshcollider reviewed (commented):
utACK 80dc829be7f8c3914074b85bb4c125baba18cb2c

SMALLjohnson commented:
3bf162f8cedac44a6b91c930fle4a866b66c7d64412add45088f512634aa80ae

fanquake commented:
Added to #23276 for backport to 22.x

MarcoFalke commented on src/test/transaction_tests.cpp:
```diff
@@ -810,10 +810,10 @@ BOOST_AUTO_TEST_CASE(test_IsStandard)
     CheckIsNotStandard(t, "dust");
     // not dust:
-    t.vout[0].nValue = 673;
+    t.vout[0].nValue = 674;
```
Looks like this changes behaviour between releases if the user changed the default dust rate.

MarcoFalke replied:
Same for `-blockmintxfee`, and `-incrementalRelayFee`, `sendrawtransaction`/`testmempoolaccept` feerate RPC arg, checks in `CheckFeeRate`.
